core/launch.go

```diff
@@ -0,0 +1,225 @@
+package core
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"chain/core/accesstoken"
+	"chain/core/account"
+	"chain/core/asset"
+	"chain/core/config"
+	"chain/core/fetch"
+	"chain/core/generator"
+	"chain/core/leader"
+	"chain/core/pin"
+	"chain/core/query"
+	"chain/core/rpc"
+	"chain/core/txbuilder"
+	"chain/core/txdb"
+	"chain/core/txfeed"
+	"chain/database/pg"
+	"chain/log"
+	"chain/protocol"
+	"chain/protocol/bc"
+)
+
+const (
+	blockPeriod              = time.Second
+	expireReservationsPeriod = time.Second
+)
+
+// LaunchOption describes a runtime configuration option.
+type LaunchOption func(*API)
+
+// AlternateAuth configures the Core to use authFn to authenticate
+// incoming requests in addition to the default access token authentication.
+func AlternateAuth(authFn func(*http.Request) bool) LaunchOption {
+	return func(a *API) { a.altAuth = authFn }
+}
+
+// BlockSigner configures the Core to use signFn to handle block-signing
+// requests. In production, this will be a function to call out to signerd
+// and its HSM. In development, it'll use the MockHSM.
+func BlockSigner(signFn func(context.Context, *bc.Block) ([]byte, error)) LaunchOption {
+	return func(a *API) { a.signer = signFn }
+}
+
+// IndexTransactions configures whether or not transactions should be
+// annotated and indexed for the query engine.
+func IndexTransactions(b bool) LaunchOption {
+	return func(a *API) { a.indexTxs = b }
+}
+
+// RateLimit adds a rate-limiting restriction, using keyFn to extract the
+// key to rate limit on. It will allow up to burst requests in the bucket
+// and will refill the bucket at perSecond tokens per second.
+func RateLimit(keyFn func(*http.Request) string, burst, perSecond int) LaunchOption {
+	return func(a *API) {
+		a.requestLimits = append(a.requestLimits, requestLimit{
+			key:       keyFn,
+			burst:     burst,
+			perSecond: perSecond,
+		})
+	}
+}
+
+// LocalGenerator configures the launched Core to run as a Generator.
+func LocalGenerator(gen *generator.Generator) LaunchOption {
+	return func(a *API) {
+		a.generator = gen
+		a.submitter = gen
+	}
+}
+
+// RemoteGenerator configures the launched Core to fetch blocks from
+// the provided remote generator.
+func RemoteGenerator(client *rpc.Client) LaunchOption {
+	return func(a *API) {
+		a.remoteGenerator = client
+		a.submitter = &txbuilder.RemoteGenerator{Peer: client}
```

Comment:
Maybe LocalGenerator and RemoteGenerator should each clear out the config from the other? So generator = nil here and remoteGenerator = nil there. That would defensively prevent getting into a weird inconsistent state in case we somehow applied both LaunchOptions. Orrrr, maybe we should check for that case here and panic immediately.

Maybe a generalization of the principle is that options should prevent the *API from ever getting into an inconsistent state (including starting out in an inconsistent state, as with AlternateAuth above). Whether by clobbering each other or by detecting and panicking.